scyllacloud/hostdialer_test.go

```diff
@@ -0,0 +1,194 @@
+	"context"
+	"crypto/rand"
+	"crypto/rsa"
+	"crypto/tls"
+	"crypto/x509"
+	"crypto/x509/pkix"
+	"encoding/pem"
+	"fmt"
+	"math/big"
+	"net"
+	"net/http/httptest"
+	"testing"
+	"time"
+
+	"github.com/gocql/gocql"
+)
+
+func TestHostSNIDialer_ServerNameIdentifiers(t *testing.T) {
+	connectionStateCh := make(chan tls.ConnectionState, 1)
+
+	server := httptest.NewUnstartedServer(nil)
+	server.TLS = &tls.Config{
+		VerifyConnection: func(state tls.ConnectionState) error {
+			connectionStateCh <- state
+			return nil
+		},
+	}
+	server.StartTLS()
+	defer server.Close()
+
+	clientCert, clientKey, err := generateClientCert(server)
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	certPem, err := encodeCertificates(clientCert)
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	privateKeyPem, err := encodePrivateKey(clientKey)
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	caCert, err := encodeCertificates(server.Certificate())
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	dialer := &gocql.ScyllaShardAwareDialer{Dialer: net.Dialer{}}
+	connConfig := &ConnectionConfig{
+		Datacenters: map[string]*Datacenter{
+			"us-east-1": {
+				CertificateAuthorityData: caCert,
+				Server:                   server.Listener.Addr().String(),
+				NodeDomain:               "node.domain.com",
+				// Test TLS server serving cert is signed for example.com domain,
+				// we skip verification here since we don't care about it in the test.
+				InsecureSkipTLSVerify: true,
+			},
+		},
+		AuthInfos: map[string]*AuthInfo{
+			"admin": {
+				ClientCertificateData: certPem,
+				ClientKeyData:         privateKeyPem,
+			},
+		},
+		Contexts: map[string]*Context{
+			"default": {
+				DatacenterName: "us-east-1",
+				AuthInfoName:   "admin",
+			},
+		},
+		CurrentContext: "default",
+	}
+
+	tt := []struct {
+		name        string
+		hostInfo    func() *gocql.HostInfo
+		expectedSNI string
+	}{
+		{
+			name: "node domain as SNI when host info is unknown",
+			hostInfo: func() *gocql.HostInfo {
+				return &gocql.HostInfo{}
+			},
+			expectedSNI: "node.domain.com",
+		},
+		{
+			name: "host SNI when host is known",
+			hostInfo: func() *gocql.HostInfo {
+				hi := &gocql.HostInfo{}
+				hi.SetHostID("host-1-uuid")
+				hi.SetDatacenter("us-east-1")
+				return hi
+			},
+			expectedSNI: "host-1-uuid.node.domain.com",
+		},
+	}
+
+	for i := range tt {
+		tc := tt[i]
+		t.Run(tc.name, func(t *testing.T) {
+			ctx, cancel := context.WithCancel(context.Background())
+			defer cancel()
+
+			hostDialer := NewSniHostDialer(connConfig, dialer)
+
+			_, err = hostDialer.DialHost(ctx, tc.hostInfo())
+			if err != nil {
+				t.Fatal(err)
+			}
+
+			receivedState := <-connectionStateCh
+			if receivedState.ServerName != tc.expectedSNI {
+				t.Errorf("expected %s SNI, got %s", tc.expectedSNI, receivedState.ServerName)
+			}
+		})
+	}
+}
+
+func generateClientCert(server *httptest.Server) (*x509.Certificate, *rsa.PrivateKey, error) {
+	now := time.Now()
+
+	privateKey, err := rsa.GenerateKey(rand.Reader, 4096)
```

Comment:
I'd probably go lower for a test to reduce the generation time

Reply:
done